File: code/0015. 3Sum/solution2.java
class Solution {
    private void twoSum(int[] nums, int target, int start, List<List<Integer>> result) {
        int n = nums.length;
        Set<Integer> visited = new HashSet<>();
        
        for (int i = start; i < n; i++) {
            int num = nums[i];
            if (visited.contains(-target - num)) {
                result.add(Arrays.asList(target, -target - num, num));
                while (i + 1 < n && nums[i + 1] == num) {
                    i++;
                }
            }
            else {
                visited.add(num);
            }
        }
    }
    
    public List<List<Integer>> threeSum(int[] nums) {
        int n = nums.length;
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        
        for (int i = 0; i < n; i++) {
            if (i == 0 || nums[i] != nums[i - 1]) {
                int target = nums[i];
                twoSum(nums, target, i + 1, result);
            }
        }
        
        return result;
    }
}

File: code/0259. 3Sum Smaller/solution2.cpp
class Solution {
public:
    int threeSumSmaller(vector<int>& nums, int target) {
        int n = nums.size(), result = 0;
        sort(nums.begin(), nums.end());
        
        for (int i = 0; i < n; i++) {
            int num = nums[i];
            int p1 = i + 1, p2 = n - 1;
            while (p1 < p2) {
                if (num + nums[p1] + nums[p2] < target) {
                    result += p2 - p1;
                    p1++;
                }
                else {
                    p2--;
                }
            }
        }
        
        return result;
    }
};